ibis/common/deferred.py

```diff
@@ -0,0 +1,574 @@
+        assert not isinstance(func, (Deferred, Builder))
+        super().__init__(func=func, arg=builder(arg))
+
+    def __repr__(self):
+        try:
+            symbol = _operator_symbols[self.func]
+            return f"{symbol}{self.arg!r}"
+        except KeyError:
+            return f"{self.func.__name__}(self.arg!r)"
+
+    def build(self, context):
+        arg = self.arg.build(context)
+        return self.func(arg)
+
+
+class BinaryCall(Slotted, Builder):
+    __slots__ = ("func", "left", "right")
+    func: Callable
+    left: Builder
+    right: Builder
+
+    def __init__(self, func, left, right):
+        assert not isinstance(func, (Deferred, Builder))
+        super().__init__(func=func, left=builder(left), right=builder(right))
+
+    def __repr__(self):
+        try:
+            symbol = _operator_symbols[self.func]
+            return f"({self.left!r} {symbol} {self.right!r})"
+        except KeyError:
+            return f"{self.func.__name__}({self.left!r}, {self.right!r})"
+
+    def build(self, context):
+        left = self.left.build(context)
+        right = self.right.build(context)
+        return self.func(left, right)
+
+
+class Mapping(Slotted, Builder):
+    __slots__ = ("typ", "builders")
+
+    def __init__(self, builders):
+        typ = type(builders)
+        builders = FrozenDict({k: builder(v) for k, v in builders.items()})
+        super().__init__(typ=typ, builders=builders)
+
+    def __repr__(self):
+        items = ", ".join(f"{k!r}: {v!r}" for k, v in self.builders.items())
+        if self.typ is dict:
+            return f"{{{items}}}"
+        else:
+            return f"{self.typ.__name__}({{{items}}})"
+
+    def build(self, context):
+        items = {k: v.build(context) for k, v in self.builders.items()}
+        return self.typ(items)
+
+
+class Sequence(Slotted, Builder):
+    __slots__ = ("typ", "builders")
+    typ: type
+
+    def __init__(self, builders):
+        typ = type(builders)
+        builders = tuple(map(builder, builders))
+        super().__init__(typ=typ, builders=builders)
+
+    def __repr__(self):
+        elems = ", ".join(map(repr, self.builders))
+        if self.typ is tuple:
+            return f"({elems})"
+        elif self.typ is list:
+            return f"[{elems}]"
+        else:
+            return f"{self.typ.__name__}({elems})"
+
+    def build(self, context):
+        return self.typ(b.build(context) for b in self.builders)
+
+
+def builder(obj):
+    if isinstance(obj, Deferred):
+        return obj.__builder__
+    elif isinstance(obj, Builder):
+        return obj
+    elif isinstance(obj, collections.abc.Mapping):
+        # allow nesting deferred patterns in dicts
+        return Mapping(obj)
+    elif is_iterable(obj):
+        # allow nesting deferred patterns in tuples/lists
+        return Sequence(obj)
+    elif callable(obj):
+        return Factory(obj)
+    else:
+        # the object is used as a constant value
+        return Just(obj)
```

Comment:
Actually they are treated differently.  `Just(sequence)` won't try to create `builders` for each element of the sequence, it will "just" return the sequence as is (of course in an immutable form at the moment). 

On the other hand `builder(sequence)` will turn each element into a separate builder supporting nested builders.